eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/AsyncPaymentTriggerer.scala

```diff
@@ -0,0 +1,142 @@
+import akka.actor.typed.{ActorRef, Behavior, SupervisorStrategy}
+import fr.acinq.bitcoin.scalacompat.ByteVector32
+import fr.acinq.bitcoin.scalacompat.Crypto.PublicKey
+import fr.acinq.eclair.Logs.LogCategory
+import fr.acinq.eclair.blockchain.CurrentBlockHeight
+import fr.acinq.eclair.io.PeerReadyNotifier.NotifyWhenPeerReady
+import fr.acinq.eclair.io.{PeerReadyNotifier, Switchboard}
+import fr.acinq.eclair.payment.relay.AsyncPaymentTriggerer.Command
+import fr.acinq.eclair.{BlockHeight, Logs}
+
+/**
+ * This actor waits for an async payment receiver to become ready to receive a payment or for a block timeout to expire.
+ * If the receiver of the payment is a connected peer, spawn a PeerReadyNotifier actor.
+ */
+object AsyncPaymentTriggerer {
+  // @formatter:off
+  sealed trait Command
+  case class Start(switchboard: ActorRef[Switchboard.GetPeerInfo]) extends Command
+  case class Watch(replyTo: ActorRef[Result], remoteNodeId: PublicKey, paymentHash: ByteVector32, timeout: BlockHeight) extends Command
+  case class Cancel(paymentHash: ByteVector32) extends Command
+  private[relay] case class NotifierStopped(remoteNodeId: PublicKey) extends Command
+  private case class WrappedPeerReadyResult(result: PeerReadyNotifier.Result) extends Command
+  private case class WrappedCurrentBlockHeight(currentBlockHeight: CurrentBlockHeight) extends Command
+
+  sealed trait Result
+  case object AsyncPaymentTriggered extends Result
+  case object AsyncPaymentTimeout extends Result
+  case object AsyncPaymentCanceled extends Result
+  // @formatter:on
+
+  def apply(): Behavior[Command] = Behaviors.setup { context =>
+    Behaviors.withMdc(Logs.mdc(category_opt = Some(LogCategory.PAYMENT))) {
+      Behaviors.receiveMessagePartial {
+        case Start(switchboard) => new AsyncPaymentTriggerer(switchboard, context).start()
+      }
+    }
+  }
+}
+
+private class AsyncPaymentTriggerer(switchboard: ActorRef[Switchboard.GetPeerInfo], context: ActorContext[Command]) {
+
+  import AsyncPaymentTriggerer._
+
+  case class Payment(replyTo: ActorRef[Result], timeout: BlockHeight, paymentHash: ByteVector32) {
+    def expired(currentBlockHeight: BlockHeight): Boolean = timeout <= currentBlockHeight
+  }
+
+  case class PeerPayments(notifier: ActorRef[PeerReadyNotifier.Command], pendingPayments: Set[Payment]) {
+    def update(currentBlockHeight: BlockHeight): Option[PeerPayments] = {
+      val expiredPayments = pendingPayments.filter(_.expired(currentBlockHeight))
+      expiredPayments.foreach(e => e.replyTo ! AsyncPaymentTimeout)
+      updatePaymentsOrStop(pendingPayments.removedAll(expiredPayments))
+    }
+
+    def cancel(paymentHash: ByteVector32): Option[PeerPayments] = {
+      val canceledPayment = pendingPayments.find(_.paymentHash == paymentHash)
+      if (canceledPayment.isDefined) canceledPayment.get.replyTo ! AsyncPaymentCanceled
+      updatePaymentsOrStop(pendingPayments.removedAll(canceledPayment))
```

Comment:
There may be multiple payments for the same payment hash, you should use `filter` instead of `find`:

```suggestion
      val canceledPayments = pendingPayments.filter(_.paymentHash == paymentHash)
      canceledPayments.foreach(_.replyTo ! AsyncPaymentCanceled)
      updatePaymentsOrStop(pendingPayments.removedAll(canceledPayments))
```

It's worth adding a unit test for this scenario.

Reply:
Oh, I wrongly assumed there could be only one. Fixed and added a test here: [9619b6e](https://github.com/ACINQ/eclair/pull/2491/commits/9619b6e3ee7c59cdfd93d3630dcce7a55d54c567)